### PR TITLE
[SPARK-50010][SQL] Expand implicit collation mismatch error

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
Include the implicit string collations in the `COLLATION_MISMATCH.IMPLICIT` error message.

### Why are the changes needed?
Make the implicit collation mismatch error more user-friendly.

### Does this PR introduce _any_ user-facing change?
Yes, implicit collation mismatch error is changed.

### How was this patch tested?
Updated existing tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by uros-db — COMMENTED** [review]
@mihailom-db please review

cc @stefankandic

**Review by stefankandic — APPROVED** [review]
LGTM!

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala [hunk 1]
```diff
@@ -191,7 +191,9 @@ object CollationTypeCasts extends TypeCoercionRule {
           .distinct
 
         if (implicitTypes.length > 1) {
-          throw QueryCompilationErrors.implicitCollationMismatchError()
+          throw QueryCompilationErrors.implicitCollationMismatchError(
+            implicitTypes.map(t => StringType(t).typeName)
```
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:195]: given that this is how it's currently done for `explicitCollationMismatchError` (code above this), how about we keep this PR as it is (given the scope)

then I propose to follow up later to correct both explicit & implicit QueryCompilationErrors like you suggested

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -3646,10 +3646,12 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
     )
   }
 
-  def implicitCollationMismatchError(): Throwable = {
+  def implicitCollationMismatchError(implicitTypes: Seq[String]): Throwable = {
     new AnalysisException(
       errorClass = "COLLATION_MISMATCH.IMPLICIT",
-      messageParameters = Map.empty
+      messageParameters = Map(
+        "implicitTypes" -> implicitTypes.map(toSQLId).mkString(", ")
```
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:3653]: given that this is how it's currently done for `explicitCollationMismatchError` (code above this), how about we keep this PR as it is (given the scope)

then I propose to follow up later to correct both explicit & implicit QueryCompilationErrors like you suggested

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:195]: the later could not happen. let's do in proper way here in the PR

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala]: fixed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala]: fixed

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 3]
```diff
@@ -3646,10 +3646,12 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
     )
   }
 
-  def implicitCollationMismatchError(): Throwable = {
+  def implicitCollationMismatchError(implicitTypes: Seq[AbstractDataType]): Throwable = {
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:3649]: Why AbstractDataType? Shouldn't we always get StringTypes?

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM, just a small nit comment

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 3 above)
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:3649]: sure, we can limit to StringType only here

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @uros-db and @stefankandic @mihailom-db for review.

**Pull request closed without merging**